### PR TITLE
Fix secondary synchronous components in main CC not computed in SA

### PR DESCRIPTION
**Pull request by obrix**
**Please check if the PR fulfills these requirements**

- [x] The commit message follows our guidelines
- [x] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)

**Does this PR already have an issue describing the problem?**

No

**What kind of change does this PR introduce?**

Bugfix
Handle properly secondary synchronous component from main connected component in SA

**What is the current behavior?**

Secondary synchronous component of main CC were ignored by the network selection on which we computed sa.

**What is the new behavior (if this is a feature change)?**

All component are now handled correctly and in a way similar as the load flow 👍 

MAIN component mode will compute SA on main connected and all its synchronous component.
ALL component mode will compute SA on all connected component and all synchronous component. 

**Does this PR introduce a breaking change or deprecate an API?**
- [x] No

### CODE REVIEW

**Review by vidaldid-rte — APPROVED** [review]

File: src/main/java/com/powsybl/openloadflow/sa/AbstractSecurityAnalysis.java [hunk 1]
```diff
@@ -262,38 +261,56 @@ SecurityAnalysisResult runSimulationsOnAllComponents(LfNetworkList networks, Lis
...
+    }
+
+    static boolean isMainCCandMainSC(LfNetwork n) {
+        return n.getNumCC() == ComponentConstants.MAIN_NUM && n.getNumSC() == ComponentConstants.MAIN_NUM;
+    }
+
+    static boolean isMainCCandSecondarySC(LfNetwork n) {
+        return n.getNumCC() == ComponentConstants.MAIN_NUM && n.getNumSC() != ComponentConstants.MAIN_NUM;
+    }
+
```
Comment (jeandemanged) [inline src/main/java/com/powsybl/openloadflow/sa/AbstractSecurityAnalysis.java:313]: `AbstractSecurityAnalysis` doesn't seem to be the right place for these methods, move them to `LfNetworkList` ?

File: src/main/java/com/powsybl/openloadflow/sa/AbstractSecurityAnalysis.java [hunk 2]
```diff
@@ -262,38 +261,56 @@ SecurityAnalysisResult runSimulationsOnAllComponents(LfNetworkList networks, Lis
...
         PreContingencyResult mergedPrecontingencyResult =
                 new PreContingencyResult(result.getPreContingencyResult().getStatus(),
                         new LimitViolationsResult(preContingenctViolations),
                         mergedPreContingencyNetworkResult);
         return new SecurityAnalysisResult(mergedPrecontingencyResult, postContingencyResults, operatorStrategyResults);
     }
 
+    static Optional<LfNetwork> selectValidMainComponent(LfNetworkList networks) {
+        return networks.getList().stream().filter(n -> isMainCCandMainSC(n) && n.getValidity().equals(LfNetwork.Validity.VALID)).findFirst();
+    }
```
Comment (jeandemanged) [inline src/main/java/com/powsybl/openloadflow/sa/AbstractSecurityAnalysis.java:297]: method name `selectValidMainComponent` is a confusing since we don't know if this means main CC or main SC.

It would be clearer with one method `List<LfNetwork> getNetworksToSimulate(LoadFlowParameters.ConnectedComponentMode mode)`

- in mode MAIN: returns all valid CC0 SCx
- in mode ALL: retuens all valid CCx SCy

LfNetworkList would be the right place for this method.

File: src/test/java/com/powsybl/openloadflow/sa/OpenSecurityAnalysisTest.java
Comment (jeandemanged) [inline src/test/java/com/powsybl/openloadflow/sa/OpenSecurityAnalysisTest.java:1]: The new tests should actually run an SA and check what results are present/absent depending on connected component mode.

**Review by jeandemanged — CHANGES_REQUESTED** [review]

File: src/main/java/com/powsybl/openloadflow/sa/AbstractSecurityAnalysis.java (hunk 2 above)
Comment (obrix) [reply src/main/java/com/powsybl/openloadflow/sa/AbstractSecurityAnalysis.java:297]: That is not what needs to be done here. I understand your approach but : 
We need a way here to find a "single" main component (usually CC0|SC0) to initiate the result structure. Here your MAIN mode would yield a list of several component on which we should again select the first one.

In fact thos methods are highly linked to the security analysis flow and only are this way because of the constraints on merged results building. 
This is also why I kept that in AbstractSecurityAnalysis, this is not supposed to be used elsewhere and will probably disappear when we rethink a more suitable structure for merged result someday.

File: src/test/java/com/powsybl/openloadflow/sa/OpenSecurityAnalysisTest.java
Comment (obrix) [reply src/test/java/com/powsybl/openloadflow/sa/OpenSecurityAnalysisTest.java:1]: I think having unit tests here is a good thing, and there is already test running full SA with several components.
I can add a full test with a SA for the case that was not handled though

File: src/main/java/com/powsybl/openloadflow/sa/AbstractSecurityAnalysis.java (hunk 2 above)
Comment (jeandemanged) [reply src/main/java/com/powsybl/openloadflow/sa/AbstractSecurityAnalysis.java:297]: I think this is exactly what needs to be done. Expectation is:
- in mode MAIN: runs all valid CC0 SCx
- in mode ALL: runs all valid CCx SCy
But the code is much more complicated, only to accommodate technical issues.
Also there is the assumption that CC0 SC0 always exists, which is as you wrote "usually" true / "usually" works, but can be wrong, I added a test case to demonstrate this.
Could be done later on if you prefer, but I find the code hard to follow.

**Review by jeandemanged — APPROVED** [review]
ty

File: src/main/java/com/powsybl/openloadflow/sa/AbstractSecurityAnalysis.java [hunk 3]
```diff
@@ -251,49 +251,64 @@ SecurityAnalysisReport runSync(SecurityAnalysisParameters securityAnalysisParame
...
-                new PreContingencyResult(result.getPreContingencyResult().getStatus(),
-                        new LimitViolationsResult(preContingenctViolations),
-                        mergedPreContingencyNetworkResult);
+            new PreContingencyResult(result.getPreContingencyResult().getStatus(),
+                new LimitViolationsResult(preContingencyViolations),
+                mergedPreContingencyNetworkResult);
         return new SecurityAnalysisResult(mergedPrecontingencyResult, postContingencyResults, operatorStrategyResults);
     }
 
+    static List<LfNetwork> getNetworkToSimulate(LfNetworkList networks, LoadFlowParameters.ConnectedComponentMode mode) {
```
Comment (jeandemanged) [inline src/main/java/com/powsybl/openloadflow/sa/AbstractSecurityAnalysis.java:299]:
```suggestion
    static List<LfNetwork> getNetworksToSimulate(LfNetworkList networks, LoadFlowParameters.ConnectedComponentMode mode) {
```

**Review by jeandemanged — CHANGES_REQUESTED** [review]
sorry one more small change please

**Review by jeandemanged — APPROVED** [review]
